components/cdc/src/endpoint.rs

```diff
@@ -120,25 +142,341 @@ impl Endpoint {
         info!("cdc register region"; "region_id" => region_id);
         unimplemented!()
     }
+
+    pub fn on_incremental_scan(
+        &mut self,
+        _region_id: u64,
+        _downstream_id: DownstreamID,
+        _entries: Vec<Option<TxnEntry>>,
+    ) {
+        unimplemented!();
+    }
+
+    fn on_region_ready(&mut self, _region_id: u64, _resolver: Resolver, _region: Region) {
+        unimplemented!();
+    }
+
+    fn on_min_ts(&mut self, _min_ts: TimeStamp) {
+        unimplemented!();
+    }
+}
+
+struct Initializer {
+    workers: PoolSender,
+    sched: Scheduler<Task>,
+
+    region_id: u64,
+    downstream_id: DownstreamID,
+    checkpoint_ts: TimeStamp,
+    batch_size: usize,
+
+    build_resolver: bool,
+}
+
+impl Initializer {
+    fn on_change_cmd(&self, mut resp: ReadResponse<RocksEngine>) {
+        if let Some(region_snapshot) = resp.snapshot {
+            assert_eq!(self.region_id, region_snapshot.get_region().get_id());
+            let region = region_snapshot.get_region().clone();
+            self.async_incremental_scan(region_snapshot, region);
+        } else {
+            assert!(
+                resp.response.get_header().has_error(),
+                "no snapshot and no error? {:?}",
+                resp.response
+            );
+            let err = resp.response.take_header().take_error();
+            let deregister = Task::Deregister {
+                region_id: self.region_id,
+                downstream_id: Some(self.downstream_id),
+                err: Some(Error::Request(err)),
+            };
+            self.sched.schedule(deregister).unwrap();
+        }
+    }
+
+    fn async_incremental_scan<S: Snapshot + 'static>(&self, snap: S, region: Region) {
+        let downstream_id = self.downstream_id;
+        let sched = self.sched.clone();
+        let batch_size = self.batch_size;
+        let checkpoint_ts = self.checkpoint_ts;
+        let build_resolver = self.build_resolver;
+        info!("async incremental scan";
+            "region_id" => region.get_id(),
+            "downstream_id" => ?downstream_id);
+
+        // spawn the task to a thread pool.
+        let region_id = region.get_id();
+        self.workers
+            .spawn(lazy(move || {
+                let mut resolver = if build_resolver {
+                    Some(Resolver::new())
+                } else {
+                    None
+                };
+
+                // Time range: (checkpoint_ts, current]
+                let current = TimeStamp::max();
+                let mut scanner = ScannerBuilder::new(snap, current, false)
+                    .range(None, None)
+                    .build_delta_scanner(checkpoint_ts)
+                    .unwrap();
+                let mut done = false;
+                while !done {
+                    let entries =
+                        match Self::scan_batch(&mut scanner, batch_size, resolver.as_mut()) {
```

Comment:
Can we use `TxnEntryScanner::scan_entries`?

Reply:
The current implementation in CDC branch needs `Vec<Option<TxnEntry>>` here, and a `None` is used to indicate that scanning has finished. If we use scan_entries here it seems this behavior is better to be changed. Is that ok?

Reply:
An empty `EntryBatch` means finished. I don't think there is a big difference in behavior.

Reply:
ok

Reply:
PTAL again. I kept the scan_batch function to group scanning and resolver building together